File: sprint3/By languages/go/lesson12code1.go
package main

import (
    "math/rand"
)

func partition(array []int, pivot int) ([]int, []int, []int) {
    var left []int
    var center []int
    var right []int

    for _, x := range array {
        if x < pivot {
            left = append(left, x)
        } else if x == pivot {
            center = append(center, x)
        } else {
            right = append(right, x)
        }
    }

    return left, center, right
}

func quicksort(array []int) []int {
    if len(array) < 2 {
        return array
    } else {
        pivot := array[rand.Intn(len(array))]
        left, center, right := partition(array, pivot)
        return append(append(quicksort(left), center...), quicksort(right)...)
    }
}

func main() {
    array := []int{5, 2, 7, 3, 9, 1, 8, 6, 4}
    sorted := quicksort(array)
    fmt.Println(sorted)
}

File: sprint3/By lessons/lesson12code1/lesson12code1.java
import java.util.ArrayList;
import java.util.List;
import java.util.Random;

public class QuickSort {
    public static List<Integer>[] partition(List<Integer> array, int pivot) {
        List<Integer> left = new ArrayList<>();
        List<Integer> center = new ArrayList<>();
        List<Integer> right = new ArrayList<>();
        for (Integer x : array) {
            if (x < pivot) {
                left.add(x);
            } else if (x == pivot) {
                center.add(x);
            } else {
                right.add(x);
            }
        }
        return new List[]{left, center, right};
    }

    public static List<Integer> quicksort(List<Integer> array) {
        if (array.size() < 2) {
            return array; // массивы с 0 или 1 элементами фактически отсортированы
        } else {
            Random random = new Random();
            int pivot = array.get(random.nextInt(array.size()));
            List<Integer>[] parts = partition(array, pivot);
            return concatenate(quicksort(parts[0]), parts[1], quicksort(parts[2]));
        }
    }

    public static List<Integer> concatenate(List<Integer> a, List<Integer> b, List<Integer> c) {
        List<Integer> result = new ArrayList<>(a.size() + b.size() + c.size());
        result.addAll(a);
        result.addAll(b);
        result.addAll(c);
        return result;
    }
}
